### ISSUE TITLE
Move functions in functions.php to a class

### ISSUE THREAD

**Issue by JoyceBabu** [mongodb/mongo-php-library#319]
Is it possible to move the functions in `functions.php` to a `Helper` or `Util` class, so that the code is autoloaded? Is there any specific reason (better performance?) for choosing functions over a `static` class?

I am having a custom autoloader that also loads composer classes. Unfortunately, as the documentation states, it is not possible to autoload the `functions.php` file. Having to manually include the file is not an elegant solution IMHO.

**Comment by Ocramius** [comment]
That's *the* solution for userland functions in PHP, sorry :-\

There's no real added value in having the autoloader kick in: if opcache is on, always including the file or calling the autoloader function pretty much has the same impact all the time.

**Comment by jmikola** [comment]
I'm ambivalent about grouping our utility functions into a common class, as they have no relation to each other. Long before autoloading, utility classes were mainly a way to keep functions out of the global namespace, which we can now do with namespaces.

Other libraries using a `functions.php` file and namespaced functions include the [AWS toolkit](https://github.com/aws/aws-sdk-php/blob/master/src/functions.php), [Guzzle](https://github.com/guzzle/guzzle/blob/master/src/functions.php), and [React PHP](https://github.com/reactphp/promise/blob/master/src/functions.php).

That said, there have been issues in other projects with `functions.php` and redeclaration errors, as Composer does not use `require_once` to include these files. Related issues in other projects:

 * https://github.com/reactphp/promise/issues/23
 * https://github.com/guzzle/guzzle/pull/1146

We've yet to have any reports about this for the library (I assume folks aren't installing it as a global package), but we should probably fix this preemptively.

----

On a side note: there is an RFC for [function autoloading](https://wiki.php.net/rfc/function_autoloading) in PHP core; however, it's been sitting in limbo for a couple of years.

**Comment by JoyceBabu** [comment]
@Ocramius The performance advantage may be negligible, but it has the additional advantage that you don't have to type the `include` statement at the beginning of every file where MongoDB is used. Of course, you can `include` the file inside the bootstrap file, but including a file in all pages for that reason does not feel like a good solution.

> We've yet to have any reports about this for the library (I assume folks aren't installing it as a global package), but we should probably fix this preemptively.

That issue might not have happened, but there are other users who have faced issue with autoloading

https://github.com/mongodb/mongo-php-library/issues/274
https://github.com/mongodb/mongo-php-library/issues/89

If there are no major disadvantage to doing this, I kindly appeal to you to consider this request positively. It won't cause any harm, but will ease the life of some.

**Comment by Ocramius** [comment]
> you don't have to type the `include` statement at the beginning of every file where MongoDB is used.

Why not just use the composer-provided installation? Safer, easier, autoloaded.

**Comment by JoyceBabu** [comment]
It was due to performance consideration. I have a huge dependency list, most of which are not required for majority of the requests. I wanted to obtain best performance, so I am using a custom autoloader which loads my custom classes first, and loads composer autoloader only when no matches are found. Further, my custom autoloader caches the classes autoloaded from composer in RAM, so that next time it does not have to load the entire dependency array.

But to be frank, I have never benchmarked the performance of my custom autoloader with that of composer.

**Comment by Ocramius** [comment]
I suggest doing that before putting performance considerations into discussion ;-)

**Comment by petr-buchin** [comment]
@JoyceBabu , I think you did not read the Composer documentation :)
Composer does not include all of it's dependencies when you include composer autoloader (usually it's file `vendor/autoload.php` in your project root), and each class file is only included, when this class is used explicitly. 
So, most likely that your custom autoloader degrades performance of your application :)

**Comment by JoyceBabu** [comment]
Yes, I understand that. But it maintains a map of all the class names to their corresponding files. Those arrays can grow really large, depending on your dependency list.

But as @Ocramius said, I think it is better to profile the code before continuing this discussion on performance any further.

**Comment by petr-buchin** [comment]
@JoyceBabu , composer by default does **not** maintains the map of all class names to their corresponding files.

Composer stores a map of all **namespace roots** to their corresponding **directories**.
That said, if, for example, your project depends on this library, that map will contain such element in a map:

```php
[
// other namespaces,
'\MongoDB\' => '{YOUR_PROJECT_ROOT}/vendor/mongodb/mongodb/src'
]
```

And if library follows PSR-0 or PSR-4 autoload rules, any class can be included as follows (let's use class `MongoDB\Client` as an example):
1. Get namespace root for namespace `MongoDB`, which is `vendor/mongodb/mongodb/src`.
2. Append class name to that path, i.e 'vendor/mongodb/mongodb/src/Client.php'.

**Comment by Ocramius** [comment]
> @JoyceBabu , composer by default does not maintain a map of all class names to their corresponding files.

That's only for the dev mode. Regardless of size (and I've seen huge projects), `--optimize-autoloader` (which dumps a classmap) is a good choice anyway.

**Comment by petr-buchin** [comment]
@Ocramius , dumping a huge class map will improve performance, but also will slightly increase memory usage. But that's all is not the real problem.
